File: Algorithms/680. Valid Palindrome II/ValidPalindrome.java
//Author: Tushar Jaiswal
//Creation Date: 12/05/2020

/* Given a non-empty string s, you may delete at most one character. Judge whether you can make it a palindrome.

Example 1:
Input: "aba"
Output: True

Example 2:
Input: "abca"
Output: True
Explanation: You could delete the character 'c'.

Note: The string will only contain lowercase characters a-z. The maximum length of the string is 50000.*/

/*Runtime Complexity: O(n).
Space Complexity: O(1)*/

class Solution {
    public boolean validPalindrome(String s) {
        for (int i = 0, j = s.length() - 1; i < j; i++, j--) {
            if (s.charAt(i) != s.charAt(j)) {
                boolean successOnDeleteLeft = true;
                boolean successOnDeleteRight = true;
                for (int k = i + 1, l = j; k < l; k++, l--) {
                    if(s.charAt(k) != s.charAt(l)) {
                        successOnDeleteLeft = false;
                        break;
                    }
                }
                for (int k = i, l = j - 1; k < l; k++, l--) {
                    if(s.charAt(k) != s.charAt(l)) {
                        successOnDeleteRight = false;
                        break;
                    }
                }
                return successOnDeleteLeft || successOnDeleteRight ? true : false;
            }
        }
        return true;
    }
}

File: Algorithms/680. Valid Palindrome II/ValidPalindrome.py
# Author: Tushar Jaiswal
# Creation Date: 2025-08-01

# Problem Source: https://leetcode.com/problems/valid-palindrome-ii

# Given a string s, return true if the s can be palindrome after deleting at most one character from it.

# Example 1:
# Input: s = "aba"
# Output: true

# Example 2:
# Input: s = "abca"
# Output: true
# Explanation: You could delete the character 'c'.

# Example 3:
# Input: s = "abc"
# Output: false

# Runtime Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def validPalindrome(self, s: str) -> bool:
        for i in range(len(s) // 2):
            j = len(s) - 1 - i 
            if s[i] != s[j]:
                # Try removing a character from beginning and end of current substrand see if either becomes a palindrome
                # If removal didn't make it a palindrome, mismatch prevents it from being palindrome so return false
                return self.isPalindrome(s[i + 1 :j + 1]) or self.isPalindrome(s[i:j])

        # Valid palindrome as didn't need to remove anything as all characters matched from beginning to end
        return True

    def isPalindrome(self, s: str) -> bool:
        for i in range(len(s) // 2):
            j = len(s) - 1 - i 
            if(s[i] != s[j]):
                return False
        return True
